# HG changeset patch
# User Rik <octave@nomad.inbox5.com>
# Date 1337986925 25200
#      Fri May 25 16:02:05 2012 -0700
# Node ID e3da2a68d709399ac6f69aed68b168fbb8fdc15b
# Parent  22244a235fd09f9d17898508666bd2491a6aecf8
Fix memory exhausted bug with log plots of small negative constant data (bug #36493)

* graphics.cc (get_axis_limits): Widen small axis limits based on sign of smaller axis limit.

diff --git a/src/graphics.cc b/src/graphics.cc
--- a/src/graphics.cc
+++ b/src/graphics.cc
@@ -5923,18 +5923,27 @@ axes::properties::get_axis_limits (doubl
           if ((min_val <= 0 && max_val > 0))
             {
               warning ("axis: omitting non-positive data in log plot");
               min_val = min_pos;
             }
           // FIXME -- maybe this test should also be relative?
           if (std::abs (min_val - max_val) < sqrt (DBL_EPSILON))
             {
-              min_val *= 0.9;
-              max_val *= 1.1;
+              // Widen range when too small
+              if (min_val >= 0)
+                {
+                  min_val *= 0.9;
+                  max_val *= 1.1;
+                }
+              else
+                {
+                  min_val *= 1.1;
+                  max_val *= 0.9;
+                }
             }
           if (min_val > 0)
             {
               // Log plots with all positive data
               min_val = pow (10, gnulib::floor (log10 (min_val)));
               max_val = pow (10, std::ceil (log10 (max_val)));
             }
           else
